src/internal/storage/kv/filesystem.go

```diff
@@ -0,0 +1,206 @@
+package kv
+
+import (
+	"context"
+	"encoding/hex"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync/atomic"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/log"
+	"github.com/pachyderm/pachyderm/v2/src/internal/miscutil"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pacherr"
+	"github.com/pachyderm/pachyderm/v2/src/internal/stream"
+	"github.com/pachyderm/pachyderm/v2/src/internal/uuid"
+	"go.uber.org/zap"
+	"golang.org/x/sync/semaphore"
+)
+
+var _ Store = &FSStore{}
+
+type FSStore struct {
+	dir          string
+	maxKeySize   int
+	maxValueSize int
+
+	// initDone is true after the store has been initialized
+	initDone atomic.Bool
+	initSem  *semaphore.Weighted
+}
+
+func NewFSStore(dir string, maxKeySize, maxValueSize int) *FSStore {
+	return &FSStore{
+		dir:          dir,
+		maxKeySize:   maxKeySize,
+		maxValueSize: maxValueSize,
+		initSem:      semaphore.NewWeighted(1),
+	}
+}
+
+func (s *FSStore) Put(ctx context.Context, key, value []byte) (retErr error) {
+	if len(key) > s.maxKeySize {
+		return errors.Errorf("max key size %d exceeded. len(key)=%d", s.maxKeySize, len(key))
+	}
+	if len(value) > s.maxValueSize {
+		return errors.Errorf("max value size %d exceeded. len(value)=%d", s.maxKeySize, len(value))
+	}
+	if err := s.ensureInit(ctx); err != nil {
+		return err
+	}
+	staging := s.stagingPathFor(key)
+	final := s.finalPathFor(key)
+	defer s.cleanupFile(ctx, &retErr, staging)
```

Comment:
This is a clever way of handling an error from a deferred function, but won't the error get overridden if os.WriteFile or os.Rename fail? If we care enough about those errors, we could use the `multierr` package to handle combining multiple errors and append the error from `s.cleanupFile` onto a potential error.

Reply:
Nevermind, I see the way this is handled is by logging an error if `retErr` is not nil.